kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/extensions/CoroutineLoggingTestExecutionFilter.kt

```diff
@@ -0,0 +1,81 @@
+package io.kotest.engine.test.extensions
+
+import io.kotest.common.ExperimentalKotest
+import io.kotest.core.config.configuration
+import io.kotest.core.listeners.Listener
+import io.kotest.core.test.TestCase
+import io.kotest.core.test.TestContext
+import io.kotest.core.test.TestResult
+import io.kotest.engine.test.withCoroutineContext
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlin.coroutines.AbstractCoroutineContextElement
+import kotlin.coroutines.CoroutineContext
+
+@ExperimentalKotest
+interface LogListener : Listener {
+   suspend fun consume(testCase: TestCase, logs: List<Any>)
+}
+
+@ExperimentalKotest
+object ConsoleLogListener : LogListener {
+   override val name = "ConsoleLogListener"
+   override suspend fun consume(testCase: TestCase, logs: List<Any>) {
+      println(" - ${testCase.description}")
+      logs.forEach { println(it) }
+   }
+}
+
+@ExperimentalKotest
+private class TestContextLoggingCoroutineContextElement(val logs: MutableList<Any>) : AbstractCoroutineContextElement(Key) {
+   companion object Key : CoroutineContext.Key<TestContextLoggingCoroutineContextElement>
+}
+
+@ExperimentalKotest
+internal object CoroutineLoggingTestExecutionExtension : TestExecutionFilter {
+   override suspend fun execute(
+      test: suspend (TestCase, TestContext) -> TestResult
+   ): suspend (TestCase, TestContext) -> TestResult = { testCase, context ->
+      val loggingListeners = configuration.listeners().filterIsInstance<LogListener>().map { it to Mutex() }
+
+      if (configuration.logLevel.isDisabled() || loggingListeners.isEmpty()) {
+         test(testCase, context)
+      } else {
+         val contextWithLogging = context.withCoroutineContext(TestContextLoggingCoroutineContextElement(mutableListOf()))
+
+         try {
+            test(testCase, contextWithLogging)
+         } catch (ex: Exception) {
+            throw ex
+         } finally {
+            val logs = contextWithLogging.getLogs()
+            loggingListeners.forEach { (listener, mut) -> mut.withLock { runCatching { listener.consume(testCase, logs) } } }
```

Comment:
What's the mutex there to avoid ?

Reply:
I think this was an attempt to prevent multiple tests from logging at the same time - which means I need to put the mutex on the listener itself, I think.